File: 0118-pascals-triangle/0118-pascals-triangle.cpp
class Solution {
public:
    vector<vector<int>> generate(int numRows) {
        vector<vector<int>>ans(numRows);
        // Handle the absolute base case
        if (numRows == 0) return ans;

        // Build the first row
        ans[0].push_back(1); 
        if (numRows == 1) return ans; 

        // Build the second row
        ans[1].push_back(1); 
        ans[1].push_back(1); 
        if (numRows == 2) return ans; 

        
        for(int i=2;i<numRows;i++)
        {
            for(int j=0;j<=i;j++)
            {
                if(j==0 || j==i)
                {
                    ans[i].push_back(1);
                }
                else
                {
                    ans[i].push_back(ans[i-1][j]+ans[i-1][j-1]);
                }
            }
        }
        return ans;
    }
};

File: 0118-pascals-triangle/0118-pascals-triangle.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> ans = new ArrayList<>();
        
        for (int i = 0; i < numRows; i++) {
            List<Integer> row = new ArrayList<>();
            
            for (int j = 0; j <= i; j++) {
                // The first and last elements of every row are always 1
                if (j == 0 || j == i) {
                    row.add(1);
                } else {
                    // Fetch and sum the two elements directly above from the previous row
                    int left = ans.get(i - 1).get(j - 1);
                    int right = ans.get(i - 1).get(j);
                    row.add(left + right);
                }
            }
            
            ans.add(row);
        }
        
        return ans;
    }
}
